src/v/cluster/partition_balancer_planner.cc

```diff
@@ -0,0 +1,309 @@
+        unavailable_nodes.push_back(model::broker_shard{node_id, 0});
+    }
+    allocation_constraints.hard_constraints.push_back(
+      ss::make_lw_shared<hard_constraint_evaluator>(
+        distinct_from(unavailable_nodes)));
+
+    auto constraints = partition_constraints(
+      assignments.id,
+      topic_metadata.get_configuration().replication_factor,
+      allocation_constraints);
+
+    auto reallocation = _partition_allocator.local().reallocate_partition(
+      constraints, stable_assigments);
+
+    if (
+      reallocation.has_value()
+      && reallocation.value().get_assignments().size() == 1) {
+        for (const auto r :
+             reallocation.value().get_assignments().front().replicas) {
+            if (
+              std::find(stable_replicas.begin(), stable_replicas.end(), r)
+              == stable_replicas.end()) {
+            }
+            _assigned_reallocation_sizes.insert({r.node_id, 0});
+            _assigned_reallocation_sizes[r.node_id] += partition_size;
+        }
+    }
+    return reallocation;
+}
+
+void partition_balancer_planner::get_unavailable_nodes_reassigments(
+  std::vector<ntp_reassigments>& reassigments,
+  cluster_health_report& health_report) {
+    for (const auto& t : _topic_table.local().topics_map()) {
+        for (const auto& a : t.second.get_assignments()) {
+            if (_reallocation_ordered == _config.batch_size) {
+                return;
+            }
+            std::vector<model::broker_shard> stable_replicas;
+            bool need_reallocation = false;
+            for (const auto& r : a.replicas) {
+                if (!_unavailable_nodes.contains(r.node_id)) {
+                    stable_replicas.push_back(
+                      model::broker_shard{r.node_id, r.shard});
+                } else {
+                    need_reallocation = true;
+                }
+            }
+            if (need_reallocation) {
+                auto new_allocation_units = get_reallocation(
+                  a,
+                  t.second,
+                  stable_replicas,
+                  health_report,
+                  get_partition_size_from_heath_report(
+                    t.second.get_configuration().tp_ns, a.id, health_report));
+                if (
+                  !new_allocation_units
+                  || new_allocation_units.value().get_assignments().size()
+                       == 0) {
+                    vlog(
+                      clusterlog.info,
+                      "Can't reallocate {} {} {} from down node",
+                      t.first.ns,
+                      t.first.tp,
+                      a.id);
+                } else {
+                    reassigments.emplace_back(ntp_reassigments{
+                      .ntp = model::ntp(t.first.ns, t.first.tp, a.id),
+                      .allocation_units = std::move(
+                        new_allocation_units.value())});
+                    _reallocation_ordered += 1;
```

Comment:
we should use partition size here, not partition count

Reply:
fixed